### DISCUSSION TITLE
Can a UserControl contain Windows? (as children)

### DISCUSSION THREAD

**Discussion by najak3d** [AvaloniaUI/Avalonia#7630] [Q&A] [answered]
Working on Urholonia here, trying to enable an Avalonia UI within an Avalonia UI.

In our Game Editor Toolset, there is a central Game Panel (see below), which is an Avalonia UserControl.

We want to put more Avalonia Windows INSIDE of this Game Panel UserControl.  Then the UserControl would form the new bounds for the contained Windows.  As the Game Panel is resized, the contained Windows would also resize, stretch, etc, to stay proportionate.

We want this to be Windows inside of a UserControl.    Does Avalonia provide a way to do this? 

We are trying to enable our GameDevs to create multiple draggable windows within their UI, and not limit them to "only adding controls/panels".

![image](https://user-images.githubusercontent.com/10914515/154391878-804c8fd1-080e-4b79-bfdf-7881769e89ec.png)

**Comment by najak3d** [comment]
Our first attempt at this might be the following:
1. Create new base class called UrholoniaWindow that derives from Avalonia's Window class, and demands a "Container".     
2. This Window will have normalized values for Position and Size (0 to 1).
3. Even in non-Editor mode, the "Container" will be the entire Game Window itself.
4. Listens to Container.LayoutChanged event and adapt Position/Size to scale with the container, and stay inside of it.

It would be nice to have some feedback from experts on your perceived viability and pitfalls of this approach.

**Comment by najak3d** [comment]
FYI, so far, this approach seems to be working.  I only needed one interface:

```C#
	public interface IWindowContainer
	{
		public void Update();
		public Size ClientSize { get; }
		public PixelPoint Position { get; }
		public event Action<IWindowContainer> LayoutChanged;
	}
```

So I've created a default "NormalizedWindowLayoutHandler" which listens to "LayoutChanged" event, and maintains a Normalized Position/Size, and so when the Container changes size - it then uses this easy math to set the Window Position, Height, Width.

All is working great.   As I resize the UrhoWindow, the Avalonia Windows are auto-resizing as it goes, to maintain the set Normalized Position/Size settings.

Next up - a Container inside of a container.... Which is what we'll be doing for the Game Panel inside of the Game Editor.  I see no reason this won't work out 100% as intended. 

This feels clean and lean so far.  Avalonia wins again.

**Comment by timunie** [comment]
I think you already know this, but I want to link it here anyway: https://github.com/wieslawsoltes/Dock

See it in action: https://www.youtube.com/watch?v=KHH23BFgU30

Happy coding
Tim

**Reply by najak3d** [reply]
Yes, and we are loving the Dock system, and can't wait to put it to use!

What we need to accomplish for a UI inside of the Game Window, is essentially Avalonia nested inside of a Control... which means the Control needs to contain Avalonia Windows...   and I doubt Avalonia directly supports this.  So we're just augmenting the Avalonia Window with the IWindowContainer concept to programmatically change each contained window's Size/Position accordingly.

**Comment by najak3d** [comment]
Progress Update Here - The Container Method is working well.  Here's the resulting Game Editor with separate In-Game UI, that positions itself correctly overtop the Game Window, as I move/resize the Game Editor.

![image](https://user-images.githubusercontent.com/10914515/154766374-5f84bdba-0e49-46cf-936a-c70bb5e90069.png)

**Comment by najak3d** [comment, marked as answer]
In case it interests anyone, I achieved this feat with 4 very small classes (< 50 SLOCs each), as follows:

```c#
	public interface IWindowContainer
	{
		public void Update();
		public Size ClientSize { get; }
		public PixelPoint Position { get; }
		public event Action<IWindowContainer> LayoutChanged;
	}
	public class UrhoWindowContainer : IWindowContainer
	{
		private U.Application _urhoApp;
		private U.Graphics _graphics;
		private U.IntVector2 _containerSize;

		public PixelPoint Position { get; private set; }
		public Size ClientSize { get; private set; }

		public event Action<IWindowContainer> LayoutChanged;

		public UrhoWindowContainer(U.Application urhoApp)
		{
			_urhoApp = urhoApp;
			_graphics = urhoApp.Graphics;
			_UpdateLayout();

			urhoApp.Graphics.WindowPos += Graphics_WindowPos;
		}

		public void Update()
		{
			var gSize = _graphics.Size;
			if (gSize != _containerSize)
			{
				_UpdateLayout();
			}
		}

		private void _UpdateLayout()
		{
			var wPos = _urhoApp.Graphics.WindowPosition;
			Position = new PixelPoint(0, 0);// wPos.X, wPos.Y);

			_containerSize = _graphics.Size;
			ClientSize = new Size(_containerSize.X, _containerSize.Y);

			try { LayoutChanged?.Invoke(this); }
			catch { } // RRRRRRRRRRRRR- Logging
		}

		private void Graphics_WindowPos(U.WindowPosEventArgs obj)
		{
			_UpdateLayout();
		}
	}

	public abstract class _BaseLayoutHandler
	{
		public IWindowContainer Container { get; private set; }
		public AC.Window Window { get; private set; }
		private A.Size _lastWindowSize;
		private A.PixelPoint _lastWindowPos;

		public _BaseLayoutHandler(AC.Window window, IWindowContainer container)
		{
			Window = window;
			Container = container;
			Container.LayoutChanged += Container_LayoutChanged;
			Window.LayoutUpdated += Window_LayoutUpdated;
		}

		private void Window_LayoutUpdated(object sender, EventArgs e)
		{
			if ((Window.ClientSize != _lastWindowSize) || (Window.Position != _lastWindowPos))
				UpdateLayout();
		}

		private void Container_LayoutChanged(IWindowContainer container)
		{
			if (Container == container)
				UpdateLayout();
		}

		public void UpdateLayout()
		{
			_OnUpdatingLayout();

			_lastWindowPos = Window.Position;
			_lastWindowSize = new A.Size(Window.Width, Window.Height); 
		}

		abstract protected void _OnUpdatingLayout();

	}
	public class NormalizedLayoutHandler : _BaseLayoutHandler
	{
		public NormalizedLayoutHandler(AC.Window window, IWindowContainer container, U.Vector2 normPos, U.Vector2 normSize)
			: base(window, container)
		{
			_normalizedPosition = normPos;
			_normalizedSize = normSize;
		}

		public U.Vector2 NormalizedPosition
		{
			get { return _normalizedPosition; }
			set
			{
				if (_normalizedPosition == value) return;
				_normalizedPosition = value;
				UpdateLayout();
			}
		}
		private U.Vector2 _normalizedPosition;

		public U.Vector2 NormalizedSize
		{
			get { return _normalizedSize; }
			set
			{
				if (_normalizedSize == value) return;
				_normalizedSize = value;
				UpdateLayout();
			}
		}
		private U.Vector2 _normalizedSize;

		protected override void _OnUpdatingLayout()
		{	// Resize/position the window!
			var cSize = Container.ClientSize;
			var cPos = Container.Position;

			var nPos = NormalizedPosition;
			var newPos = new A.PixelPoint(cPos.X + _NormI(cSize.Width, nPos.X), cPos.Y + _NormI(cSize.Height, nPos.Y));
			if (newPos != Window.Position)
				Window.Position = newPos;

			var nSize = NormalizedSize;
			var newWidth = _NormD(cSize.Width, nSize.X);
			if (Window.Width != newWidth)
				Window.Width = newWidth;

			var newHeight = _NormD(cSize.Height, nSize.Y);
			if (Window.Height != newHeight)
				Window.Height = newHeight;
		}

		private int _NormI(double containerSpan, float normalizedValue)
		{
			int normalizedInt = (int)Math.Round(containerSpan * normalizedValue);
			return normalizedInt;
		}
		private double _NormD(double containerSpan, float normalizedValue)
		{
			return Math.Round(containerSpan * normalizedValue);
		}
	}
```

The "Container.LayoutChanged" event calls the LayoutHandler.UpdateLayout() method to do it's thing, and resize the Window in accordance with the adjustment.   It happens all same-render-frame, so it all looks clean/perfect to the user.

This is allowing us to insert a Game UI into the Game Panel itself, while user is in Game Editor mode.

**Reply by workgroupengineering** [reply]
same reply https://github.com/AvaloniaUI/Avalonia/discussions/7646#discussioncomment-2210403
